src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs

```diff
@@ -1395,9 +1395,11 @@ private void GatherTaskItemOutputs(bool outputTargetIsItem, string outputTargetN
                                     // Setting an item spec expects the escaped value, as does setting metadata.
                                     newItem = new ProjectItemInstance(_projectInstance, outputTargetName, EscapingUtilities.Escape(output.ItemSpec), parameterLocationEscaped);
 
-                                    newItem.SetMetadataOnTaskOutput(output.CloneCustomMetadata()
-                                        .Cast<DictionaryEntry>()
-                                        .Select(x => new KeyValuePair<string, string>((string)x.Key, EscapingUtilities.Escape((string)x.Value))));
+                                    newItem.SetMetadataOnTaskOutput(output.CloneCustomMetadata().Cast<object>()
+                                        .Select(x =>
+                                            x is DictionaryEntry de ? new KeyValuePair<string, string>((string)de.Key, EscapingUtilities.Escape((string)de.Value)) :
+                                            x is KeyValuePair<string, string> kvp ? new KeyValuePair<string, string>(kvp.Key, EscapingUtilities.Escape(kvp.Value)) :
+                                            throw new Exception("Metadata item was neither DictionaryEntry nor KeyValuePair<string, string>")));
```

Comment:
minor: is it worth-while just mention that the type was unexpected vs. detailing the specific types that are expected?
Any chance we'll add more valid custom data types in the future?

Reply:
Outdated - code removed